### DISCUSSION TITLE
Merge dry run option for local branches + random tip `'join the team" expression of interest

### DISCUSSION THREAD

**Discussion by jackoconnordev** [jesseduffield/lazygit#3004] [Ideas]
Hi, to start with I'd just like to say how much I enjoy using lazygit and how much of an educational experience it has been for me in terms of learning more git features.

## Idea

Sometimes before I push to a remote I want to check if a branch I'm on will have merge conflict errors with the default branch of the remote I'm working on, without actually merging in the case that there are no conflicts.

Right now I believe the current way to do this is to attempt a merge, and if successful undo the merge by pressing "z". This isn't a massiva pain by any means, but conceptually I like the idea of having a shortcut dry-run merge option which does not alter my local git files and it is a command that I currently run by exiting lazygit to have autocompletion for long branch names on the regular git command line tool, vs the custom command entrybox in lazygit.

A dry-run option inside the merge/rebase options could be a nice feature to have in terms of having a no-cost merge checking option directly in lazygit.

## Example git dry-run snippet

_Snippet source: https://gist.github.com/kewang/e97b0018237a65c6f80951bee4715a9b_
```sh
#!/bin/sh
BRANCH=$1
git merge --no-commit --no-ff $1; git status; git merge --abort
# References: https://stackoverflow.com/questions/501407/is-there-a-git-merge-dry-run-option
```

## Reference images

_Want to check if dry-run branch has merge conflicts with master_
<img width="558" alt="image" src="https://github.com/jesseduffield/lazygit/assets/134723582/119f8a84-7b1d-48a8-834e-6e7041a25a95">

_Press "m" for merge/rebase options_
<img width="954" alt="image" src="https://github.com/jesseduffield/lazygit/assets/134723582/958cbea2-c482-4b98-9979-0e417632a7db">

_Attempt dry-run could go here_
<img width="954" alt="image" src="https://github.com/jesseduffield/lazygit/assets/134723582/7e4b7759-6c59-417e-98cc-d0a9aa0e2c80">

## Join the team expression of interest

I also saw the expression of interest message in the command log while using lazygit. I'd be interested in trying to help even a little bit for both the general experience, git experience and because this is a tool I actively use.

I've recently started working full-time as a software engineer and regular open source contributions are something I'd like to work up to. I work on containerised applications which are deployed to kubernetes clusters, so I have an interest in familiarising myself with Go so that I can better understand the various controllers that the devops team have coded.

I'd also like to get experience working with a compiled language as I primarily use TypeScript, and the only other languages I have used regularly are Python and R which are also scripting languages.

_Random tip message_
<img width="1108" alt="image" src="https://github.com/jesseduffield/lazygit/assets/134723582/7c9d6caa-d0d9-4cf0-a1cc-493d98fa63b1">

**Comment by jesseduffield** [comment]
Hey @jackoconnordev glad you wanna join the team :)

The feature makes sense. I haven't tested out the bash snippet you pasted but if it can reliably test the merge without affecting the working tree then that's pretty good. We could make it so that when you press shift+m on a branch, rather than showing a confirmation prompt we show a menu with two items: one for merging and one for dry-run merging.

The merge options menu that comes up when you press 'm' shouldn't need to change.

I assume if there are no conflicts we raise an alert and tell the user that, but what happens if there are conflicts? Do we say 'there will be conflicts' or do we actually specify what they are?

**Comment by jackoconnordev** [comment]
Hi @jesseduffield, thanks for responding!

## Reliability of dry run merge

The dry run merge snippet I have works the same as a regular merge in the sense that before attempting a dry run any uncommitted changes in the local directory must either be committed or stashed. When the dry run finishes and aborts the merge with `git merge --abort` then the user will be returned to having no changes in their working directory.

_Current lazygit/git output when attempting to merge with unstashed changes_ 
<img width="975" alt="Screenshot 2023-09-12 at 09 07 00" src="https://github.com/jesseduffield/lazygit/assets/134723582/5ff21e34-948f-4c58-a753-59dc003b8d8c">

The `git merge --no-commit --no-ff` command exits with status code 1 if there are conflicts and 0 if there are no conflicts which could be an easy way to determine what to display to the user after attempting a dry run merge.

_Dry run snippet with conflict_
<img width="497" alt="Screenshot 2023-09-12 at 09 11 00" src="https://github.com/jesseduffield/lazygit/assets/134723582/6cb7e4d0-322e-446c-90b0-41dc8dc9b084">

_Dry run snippet without conflict_
<img width="499" alt="Screenshot 2023-09-12 at 09 16 16" src="https://github.com/jesseduffield/lazygit/assets/134723582/e8ee0005-d814-4214-83f3-9f5c36f14936">

## Usage within lazygit

I think the way I've been using it is just to test if there are merge conflicts or not when I'm working on several branches at the same time. In that case just an alert that there are or aren't merge conflicts would satisfy me I think.

Also, how would you feel about putting the dry run merge on \<c-m\> and that way the current merge commands/muscle memory can be left alone?

**Reply by jesseduffield** [reply]
> Also, how would you feel about putting the dry run merge on and that way the current merge commands/muscle memory can be left alone?

I'm not sure what you mean by this. Do you mean always dry run before doing a regular merge? My concern there is that it adds too much time to the action

**Reply by jackoconnordev** [reply]
Oh woops sorry I had typed \<c-m\> for control-m but hadn't realised I needed escape characters.
I've updated my previous comment to have the text display.

I've been playing around a little bit on a fork with the functionality I was describing so I'll try get something that actually runs in the next week or two if I can find the time as a mini demo.

**Reply by jesseduffield** [reply]
I would prefer just swapping out the confirmation view for a menu which has an additional option for the dry-run. We've done this loads of times in past situations because:
* muscle memory is actually preserved: in this typical case you're still pressing m + enter
* it spares ctrl+m to be used for some other purpose.
* it makes the new action (dry run) more discoverable

Let me know if you need any pointers. Looking forward to your demo :)

**Reply by jackoconnordev** [reply]
Oh sure thing, that makes a lot of sense actually 🤔.

I will do my best 🫡 and I'll ask if I get stuck at any point thanks 🙂
